File: tools/parity/dispatch/worktree.ts
// TS dispatch for the worktree parity module: maps the shared vector function
// names to the real `src/main/git/worktree.ts` exports so the harness compares
// the live TS reference against the Rust port. Only the pure porcelain parser
// is covered; the rest of the module is git IO/orchestration.

import { parseWorktreeList } from '../../../src/main/git/worktree'

export function dispatch(fn: string, input: unknown): unknown {
  switch (fn) {
    case 'parseWorktreeList': {
      const { output, nulDelimited } = input as { output: string; nulDelimited?: boolean }
      return parseWorktreeList(output, { nulDelimited })
    }
    default:
      throw new Error(`unknown function ${fn}`)
  }
}

File: rust/crates/orca-parity/src/modules/worktree.rs
//! Parity dispatch for `orca_git::worktree` vs `src/main/git/worktree.ts`.
//! Only the pure `parseWorktreeList` parser is covered; the rest of the TS
//! module is git IO/orchestration built on top of it.

use orca_git::worktree::{parse_worktree_list, worktree_list_to_json};
use serde_json::{json, Value};

pub fn dispatch(function: &str, input: &Value) -> Value {
    match function {
        "parseWorktreeList" => {
            let output = input.get("output").and_then(Value::as_str).unwrap_or("");
            let nul_delimited = input
                .get("nulDelimited")
                .and_then(Value::as_bool)
                .unwrap_or(false);
            // The napi export and this harness share one canonical serializer so
            // the production path and the parity check can't diverge.
            worktree_list_to_json(&parse_worktree_list(output, nul_delimited))
        }
        other => json!({ "__parity_error__": format!("unknown function {other}") }),
    }
}
